src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs

```diff
@@ -451,16 +459,53 @@ protected SynthesizedImplementationMethod GenerateIteratorGetEnumerator(MethodSy
                 CapturedSymbolReplacement proxy;
                 if (copyDest.TryGetValue(parameter, out proxy))
                 {
-                    bodyBuilder.Add(
-                        F.Assignment(
-                            proxy.Replacement(F.Syntax, stateMachineType => F.Local(resultVariable)),
-                            copySrc[parameter].Replacement(F.Syntax, stateMachineType => F.This())));
+                    // result.parameter = this.parameterProxy;
+                    BoundExpression left = proxy.Replacement(F.Syntax, stateMachineType => F.Local(resultVariable));
+                    BoundStatement copy = F.Assignment(
```

Comment:
Consider making this conditional on whether the method is async.

Reply:
I didn't understand. We need this statement in both cases.

---
In reply to: [277111359](https://github.com/dotnet/roslyn/pull/35121#discussion_r277111359) [](ancestors = 277111359)